### ISSUE TITLE
node_modules folder is not deployed for frontend app

### ISSUE THREAD

**Issue by cstavaru** [aws-amplify/amplify-console#829]
Hello everyone,

TL;DR: node_modules folder is skipped by the deployment process. If you rename it, it will be included. Why ?

I have an  issue with a web frontend project I'm trying to build and deploy via AWS Amplify. It's a single page application project that uses Google Polymer (including for building), however the problem is not related. 

As you may be aware, many new frontend frameworks now use npm for frontend package management, which results in a node_modules subfolder inside the public website folder. This folder is queried by the web browser for various components. Files inside this folder are altered during the build process such that they are valid when used by frontend web apps.

The build works fine, but this node_modules folder is not copied to the artifacts archive, no matter what. I list the files in the build folder and the node_modules is there, but it is never deployed. 

The repo I'm using is a monorepo. my-console/frontend is the frontend app, and the build folder is my-console/frontend/public

Any idea how to solve this ? Here is my build settings file:

version: 1
applications:
  - frontend:
      phases:
        preBuild:
          commands:
            - npm install
        build:
          commands:
            - node ./node_modules/gulp/bin/gulp.js --output-path=public
        postBuild:
          commands:
            - ls -R public
      artifacts:
        baseDirectory: public
        files:
          - '**/*'
      cache:
        paths:
          - node_modules/**/*
    appRoot: my-console/frontend

**Comment by cstavaru** [comment]
For those having issues with node_modules folder not being copied for frontend web app deployment, the solution is of course to rename it. But this impacts the build scripts which must reside in node_modules for NodeJS to be able to require them.

So you have to use a combination of npm (for scripts loaded using NodeJS for build purposes) and yarn with the --modules-path option set to the desired folder name. You can split the packages using the "dependencies" (for yarn and deployment) and "devDependencies" (for npm and build scripts) in the package.json file.

One other small issue: in the case of a Polymer/lit-html/lit-element app, you can't choose any component folder name due to some bugs in Polymer builder which does not accept arbitrary names (despite the componentDir option which is broken). But there is one option: set the --modules-path yarn option to 'bower_components' which is the older package manager supported by Polymer. It seems strange to use that name for npm packages but it works, at least until there are options to copy the node_modules folder.

**Comment by Athena96** [comment]
Hi @cstavaru can you try updating the front-end build section of your amplify.yml with the following:
```
    cache:
        paths:
            - 'node_modules/**/*'
```

**Comment by cstavaru** [comment]
> Hi @cstavaru can you try updating the front-end build section of your amplify.yml with the following:
> 
> ```
>     cache:
>         paths:
>             - 'node_modules/**/*'
> ```

Actually my posted yml has the exact path as you have suggested (which is also the default suggested by Amplify console when creating the app), without the single quotes. But Github's post editor does something when posting and alters the way it looks for readers. So that is actually the path used by my yml. I tried with an without quotes, with a different path, etc. No luck.

**Comment by hsspain** [comment]
@cstavaru Sorry for the delay. Your posted YML shows ```- node_modules//*``` and we recommend the path ```- node_modules/**/*```

Please try this and let us know if that works for you.

**Comment by difosfor** [comment]
I'm running into this issue as well. We need node_modules files to be deployed along with the other files. I don't understand why this issue was closed because renaming that directory not a solution. It would be very impractical for us. And like cstavaru said, he already configured the cache paths as you told him, it was just GitHub's Markdown converting the asterisks into bold formatting. Also, I don't see why cache would have anything to do with the artifacts files selection used to determine which files get deployed, which is where this breaks down.

Can we re-open this issue or should I create a new one? I guess I'll also check for another support channel for a faster response.

**Comment by ValYouW** [comment]
@swaminator This issue should get reopen or at least be closed with some explanation, like is it on purpose? why?

Anyway, if anyone else is looking for a workaround this is what I did:
In the build script rename `node_modules` to `nodemodules` (`fs.renameSync`) and then add a rewrite rule like:
`
/node_modules/<*> | /nodemodules/<*> | 200 (Rewrite)`
